torch/nn/modules/lazy.py

```diff
@@ -0,0 +1,243 @@
+import itertools
+from typing_extensions import Protocol
+import warnings
+
+import torch
+from ..parameter import UninitializedParameter
+
+
+class _LazyProtocol(Protocol):
+    """This is to avoid errors with mypy checks for 
+    The attributes in a mixin:
+    https://mypy.readthedocs.io/en/latest/more_types.html#mixin-classes
+    """
+    def _register_load_state_dict_pre_hook(self, hook):
+        ...
+
+    def register_forward_pre_hook(self, hook):
+        ...
+
+    def _lazy_load_hook(
+            self, state_dict, prefix, local_metadata, strict,
+            missing_keys, unexpected_keys, error_msgs):
+        ...
+
+    def _get_name(self):
+        ...
+
+    def _infer_parameters(self, module, input):
+        ...
+
+    @property
+    def _parameters(self):
+        ...
+
+    @property
+    def _buffers(self):
+        ...
+
+    @property
+    def _non_persistent_buffers_set(self):
+        ...
+
+    @property
+    def _load_hook(self):
+        ...
+
+    @property
+    def _initialize_hook(self):
+        ...
+
+
+class LazyModuleMixin:
+    r"""A mixin for modules that lazily initialize parameters.
+
+    Lazy modules are a special type of modules that can hold
```

Comment:
These updates are good but this paragraph and the next one now have redundant information. You could do something like this: 

"Modules that lazily initialize parameters, or "lazy modules," derive the shapes of their parameters from the first input(s) to their forward method. Until that first forward they contain  :class:`torch.nn.UninitializedParameter`s that should not be accessed or used, and afterward they contain regular :class:`torch.nn.Parameter`s. Lazy modules are convenient since they don't require computing some module arguments, like the `in_features` argument of a typical :class:`torch.nn.Linear`.

After construction, networks with lazy modules should first be converted to the desired dtype and placed on the desired device. The lazy modules should then be initialized with one or more "dry runs". These "dry runs" send inputs of the correct size, dtype, and device through the network and to each lazy of its lazy modules. After this the network can be used as usual.

For example:

```
class LazyMLP(torch.nn.Module):
    ...    def __init__(self):
    ...        super().__init__()
    ...        self.fc1 = torch.nn.LazyLinear(10)
    ...        self.relu1 = torch.nn.ReLU()
    ...        self.fc2 = torch.nn.LazyLinear(1)
    ...        self.relu2 = torch.nn.ReLU()
    ...
    ...    def forward(self, input):
    ...        x = self.relu1(self.fc1(input))
    ...        y = self.relu2(self.fc2(x))
    ...        return y
>>> # constructs a network with lazy modules
>>> lazy_mlp = LazyMLP()
>>> # transforms the network's device and dtype
>>> # NOTE: these transforms can and should be applied after construction and before any 'dry runs'
>>> lazy_mlp = lazy_mlp.cuda().double()  
>>> lazy_mlp
    LazyMLP(
      (fc1): LazyLinear(in_features=0, out_features=10, bias=True)
      (relu1): ReLU()
      (fc2): LazyLinear(in_features=0, out_features=1, bias=True)
      (relu2): ReLU()
    )
>>> # performs a dry run to initialize the network's lazy modules
>>> lazy_mlp(torch.ones(10, 10, dtype=torch.double, device='cuda'))  
>>> # after initialization, LazyLinear modules become regular Linear modules
>>> lazy_mlp
    MLP(
      (fc1): Linear(in_features=10, out_features=10, bias=True)
      (relu1): ReLU()
      (fc2): Linear(in_features=10, out_features=1, bias=True)
      (relu2): ReLU()
    )
>>> # attaches an optimizer, since parameters can now be used as usual
>>> optim = torch.optim.SGD(lazy_mlp.parameters(), lr=0.01)
```

Reply:
Thanks!!

Any idea of how to solve this?
```
ct 15 06:33:30 /var/lib/jenkins/workspace/docs/source/nn.rst.rst:383: WARNING: autosummary: stub file not found 'nn.lazy.LazyModuleMixin'. Check your autosummary_generate setting.
Oct 15 06:33:30 /opt/conda/lib/python3.6/site-packages/torch/nn/modules/linear.py:docstring of torch.nn.LazyLinear:36: WARNING: Explicit markup ends without a blank line; unexpected unindent.
Oct 15 06:33:30 looking for now-outdated files... none found```

I am not very good with sphinx 😅

Reply:
Updated, thanks!!!